File: backend/scaffolding-cc-main/src/main/java/scaffolding/cc/services/AutorService.java
package scaffolding.cc.services;

import org.springframework.stereotype.Service;
import scaffolding.cc.entities.AutorEntity;
import scaffolding.cc.models.Autor;

import java.util.List;

@Service
public interface AutorService {
    List<Autor> getAutores();

    Autor getAutorById(Long id);

    Autor getAutorByDni(Integer dni);

    AutorEntity saveAutor(Autor autor);

    Boolean deleteAutor(Long id);

    AutorEntity updateAutor(Integer dni, Autor autor);

    Boolean deleteByDni(Integer dni);
}

File: frontend/app-autor/src/app/services/autor.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map, Observable } from 'rxjs';
import { Autor } from '../models/Autor';

@Injectable({
  providedIn: 'root',
})
export class AutorService {
  private api = 'http://localhost:8080/api/autor';

  constructor(private http: HttpClient) {
    
  }

  saveAutor(autor: Autor): Observable<any> {
    return this.http.post<Autor>(this.api+'/saveAutor/', autor);
  }

  getAutores(): Observable<Autor[]> {
    return this.http.get<Autor[]>(this.api+'/getAutores');
  }

  updateAutor(dni: number, autor: Autor): Observable<any> {
    return this.http.put<Autor>(this.api+'/updateAutor/'+dni, autor);
  }

  deleteAutor(dni: number): Observable<any> {
    return this.http.delete<Autor>(this.api+'/deleteByDni/'+dni);
  }

  getAutorByDni(dni: number): Observable<Autor> {
    return this.http.get<Autor>(this.api+'/getAutorByDni/'+dni);
  }

  getAutorById(id: number): Observable<Autor> {
    return this.http.get<Autor>(this.api+'/getAutorById/'+id);
  }

  validarDni(dni: number): Observable<boolean> {
    return this.getAutores().pipe(
      map((autores: Autor[]) => {
        return !autores.some(autor => autor.dni === dni);
      })
    );
  }
}